File: CMPS101/pa1/Lex.java
/*
 * Name: Aaron Steele, atsteele
 * Assignment: pa1
 */

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;

public class Lex 
{
	public static void main(String[] args)
	{
		// 2 input parameters in args, input file, output file
		if(args.length != 2)
		{
			System.err.println("There are not exactly 2 arguments.");
			return;
		}
		String in = args[0];
		String out = args[1];
		
		File f = new File(in);
		if(f.exists())
		{
			String[] lines = readAllLines(in);
			int n = lines.length;
			List l = new List();
			
			
			for(int i = 0; i < n; ++i)
			{
				
				// Use insertion sort method from front to back
				l.moveFront();
				
				while(l.index() >= 0)
				{
					// if next string comes after current
					if(lines[i].compareTo(lines[l.get()]) > 0)
					{
						l.moveNext();
					}
					else // if the next string comes before or is equal to string, place before
					{
						l.insertBefore(i);
						break;
					}
				}
				
				// If cursor falls off list, append to list
				if(l.index() < 0)
					l.append(i);
				
			}
			
			
			// Write to file
			BufferedWriter outputWriter = null;
			try
			{
				outputWriter = new BufferedWriter(new FileWriter(out));
				l.moveFront();
				while(l.index() >= 0)
				{
					outputWriter.write(lines[l.get()]);
					outputWriter.newLine();
					l.moveNext();
				}
				
				outputWriter.flush();  
				outputWriter.close();
			}
			catch(Exception e)
			{
				e.printStackTrace();
			}
			System.out.println("Done. Saved to " + out + ".");
		}
		else
		{
			System.out.println("Error: The input file doesn't exist. Try again.");
		}
	}
	
	public static String[] readAllLines(String path)
	{
		Charset charset = Charset.forName("US-ASCII");
		java.util.List<String> out = new ArrayList<String>();
		
		// Taken from https://docs.oracle.com/javase/tutorial/essential/io/file.html
		try (BufferedReader reader = Files.newBufferedReader(Paths.get(path), charset)) {
		    String line = null;
		    while ((line = reader.readLine()) != null) {
		        out.add(line);
		    }
		} catch (IOException x) {
		    System.err.format("IOException: %s%n", x);
		}
		
		return out.toArray(new String[]{""});
	}
}

File: CMPS101/pa2/Lex.c
#define _CRT_SECURE_NO_DEPRECATE
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "List.h"

#define MAX_LEN 160

//int strcmp(const char*, const char*);

int main(int argc, char * argv[]) {    
    int i = 0;
    FILE *in, *out;
    char lines[MAX_LEN][BUFSIZ];


    // check command line for correct number of arguments
    if (argc != 3) {
        printf("Usage: %s <input file> <output file>\n", argv[0]);
        exit(1);
    }

    // open files for reading and writing 
    in = fopen(argv[1], "r");
    out = fopen(argv[2], "w");
    if (in == NULL) {
        printf("Unable to open file %s for reading\n", argv[1]);
        exit(1);
    }
    if (out == NULL) {
        printf("Unable to open file %s for writing\n", argv[2]);
        exit(1);
    }
    
    while (i < MAX_LEN && fgets(lines[i], sizeof(lines[0]), in))
    {
        lines[i][strlen(lines[i]) - 1] = '\0';
        i = i + 1;
    }
    
    List L = newList();

    for (int j = 0; j < i; ++j)
    {
        // Use insertion sort method from front to back
        moveFront(L);

        while (index(L) >= 0)
        {
            // if next string comes after current
            if (strcmp(lines[j], lines[get(L)]) > 0)
            {
                moveNext(L);
            }
            else // If next string comes before or is equal to string, place before
            {
                insertBefore(L, j);
                break;
            }
        }

        // If cursor falls off list, append to list
        if (index(L) < 0)
            append(L, j);
    }


    // Write to file
    moveFront(L);
    while (index(L) >= 0)
    {
        fprintf(out, "%s\n", lines[get(L)]);
        moveNext(L);
    }

    /* close files */
    fclose(in);
    fclose(out);

    printf("Done. Saved to %s.\n", argv[2]);

    return(0);
}